modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/outgoing/OutgoingSnapshot.java

```diff
@@ -133,39 +141,84 @@ public PartitionKey partitionKey() {
     }
 
     /**
-     * Freezes the scope of this snapshot.
+     * Freezes the scope of this snapshot. This includes taking snapshot metadata and opening TX data cursor.
      *
      * <p>Must be called under snapshot lock.
      */
     void freezeScope() {
         assert mvOperationsLock.isLocked() : "MV operations lock must be acquired!";
 
+        frozenMeta = takeSnapshotMeta();
+
         txDataCursor = partition.txStatePartitionStorage().scan();
     }
 
+    private SnapshotMeta takeSnapshotMeta() {
+        long lastAppliedIndex = Math.max(
+                partition.mvPartitionStorage().lastAppliedIndex(),
+                partition.txStatePartitionStorage().lastAppliedIndex()
+        );
+
+        return SnapshotMetaUtils.snapshotMetaAt(lastAppliedIndex, logManager);
+    }
+
+    /**
+     * Returns metadata corresponding to this snapshot.
+     *
+     * @return This snapshot metadata.
+     */
+    public SnapshotMeta meta() {
+        SnapshotMeta meta = frozenMeta;
+
+        assert meta != null : "No snapshot meta yet, probably the snapshot scope was not yet frozen";
+
+        return meta;
+    }
+
     /**
-     * Reads a snapshot meta and returns a future with the response.
+     * Reads the snapshot meta and returns a response. Returns {@code null} if the snapshot is already closed.
      *
-     * @param metaRequest Meta request.
+     * @param request Meta request.
      */
-    SnapshotMetaResponse handleSnapshotMetaRequest(SnapshotMetaRequest metaRequest) {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17935
+    @Nullable
+    SnapshotMetaResponse handleSnapshotMetaRequest(SnapshotMetaRequest request) {
+        if (closed) {
+            return logThatAlreadyClosedAndReturnNull();
+        }
+
+        SnapshotMeta meta = frozenMeta;
+
+        assert meta != null : "No snapshot meta yet, probably the snapshot scope was not yet frozen";
+
+        return MESSAGES_FACTORY.snapshotMetaResponse().meta(meta).build();
+    }
+
+    @Nullable
+    private <T> T logThatAlreadyClosedAndReturnNull() {
```

Comment:
What will happen to the receiver if we return `null`? will he hang?

Reply:
If we return `null`, then `OutgoingSnapshotsManager#handleSnapshotRequestMessage()` will return `null`, so we'll return no response to the receiver. I guess timeout will trigger on the receiver.

Actually, I would prefer to add an ability to return an error in all 3 response types, or even change our invocation mechanism to automatically pass exceptions from the responder to the invoker and throw them on the invoker (wrapped in something like `RemoteException`), but @ibessonov suggested to just return nothing in such cases for simplicity.